File: eater-backend/src/api/services/viewrecipedetail.service.js
const mongoose = require('mongoose');
const { Recipe } = require('../../models/Recipe');
const { RecipesIngredient } = require('../../models/recipes_ingredient');
const { RecipesStep } = require('../../models/recipes_step');
const { RecipeNutrition } = require('../../models/recipe_nutrion');
const { RecipeMicronutrientValues } = require('../../models/recipe_micronutrient_values');
const { IngredientMicronutrientValues } = require('../../models/ingredient_micronutrient_values');

class ViewRecipeDetailService {
  // Get full recipe details for meal plan view (with all nested data including ingredient micronutrients)
  async getFullRecipeDetails(recipeId) {
    try {
      console.log(`[ViewRecipeDetailService] Fetching recipe details for ID: ${recipeId}`);
      
      const recipe = await Recipe.findById(recipeId);
      if (!recipe) {
        throw new Error("Recipe not found");
      }

      // Fetch ingredients (populated)
      const ingredients = await RecipesIngredient.find({ recipeId }).populate('ingredientId');
      
      // Fetch micronutrients for each ingredient
      const ingredientsWithMicronutrients = await Promise.all(
        ingredients.map(async (ing) => {
          const micronutrients = await IngredientMicronutrientValues.find({
            ingredientId: ing.ingredientId._id
          }).populate('micronutrientId');
          
          return {
            ...ing.toObject(),
            micronutrients
          };
        })
      );

      // Fetch other recipe data
      const [steps, nutrition, recipeMicronutrients] = await Promise.all([
        RecipesStep.find({ recipeId }).sort({ stepNumber: 1 }),
        RecipeNutrition.findOne({ recipeId }),
        RecipeMicronutrientValues.find({ recipeId }).populate('micronutrientId')
      ]);

      console.log(`[ViewRecipeDetailService] Recipe fetched: ${recipe.name}, Ingredients: ${ingredientsWithMicronutrients.length}, Steps: ${steps.length}`);

      return {
        ...recipe.toObject(),
        ingredients: ingredientsWithMicronutrients,
        steps,
        nutrition,
        micronutrients: recipeMicronutrients
      };
    } catch (error) {
      console.error(`[ViewRecipeDetailService] Error fetching recipe: ${error.message}`);
      throw error;
    }
  }

  // Get recipe with basic info only
  async getRecipeBasicInfo(recipeId) {
    try {
      const recipe = await Recipe.findById(recipeId);
      if (!recipe) {
        throw new Error("Recipe not found");
      }
      return recipe;
    } catch (error) {
      throw error;
    }
  }
}

module.exports = new ViewRecipeDetailService();

File: mobile/lib/src/features/meal_plan/data/view_recipe_detail_service.dart
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';

class ViewRecipeDetailService {
  final Dio dio;

  ViewRecipeDetailService({required this.dio});

  static const String baseUrl = '/api/view-recipe-details';

  /// Get full recipe details (ingredients, steps, nutrition, micronutrients)
  /// Dedicated service for meal plan recipe viewing
  Future<Map<String, dynamic>> getFullRecipeDetails(String recipeId) async {
    try {
      debugPrint(
          '[ViewRecipeDetailService] Fetching recipe details for ID: $recipeId');

      final response = await dio.get(
        '$baseUrl/$recipeId/full-details',
      );

      if (response.statusCode == 200) {
        final data = response.data['data'] ?? response.data;
        debugPrint(
            '[ViewRecipeDetailService] Recipe details fetched successfully');
        debugPrint(
            '[ViewRecipeDetailService] Recipe: ${data['name']}, Ingredients: ${(data['ingredients'] as List?)?.length ?? 0}');

        return data as Map<String, dynamic>;
      } else {
        throw Exception(
            'Failed to fetch recipe details: ${response.statusCode}');
      }
    } on DioException catch (e) {
      debugPrint('[ViewRecipeDetailService] DioException: ${e.message}');
      throw Exception('Error fetching recipe details: ${e.message}');
    } catch (e) {
      debugPrint('[ViewRecipeDetailService] Exception: $e');
      throw Exception('Unexpected error fetching recipe details: $e');
    }
  }

  /// Get recipe basic info only
  Future<Map<String, dynamic>> getRecipeBasicInfo(String recipeId) async {
    try {
      final response = await dio.get('$baseUrl/$recipeId');

      if (response.statusCode == 200) {
        return response.data['data'] ?? response.data;
      } else {
        throw Exception('Failed to fetch recipe: ${response.statusCode}');
      }
    } on DioException catch (e) {
      throw Exception('Error fetching recipe: ${e.message}');
    } catch (e) {
      throw Exception('Unexpected error fetching recipe: $e');
    }
  }
}
